### PR TITLE
Add switch rules synchronization when switch activation.

### PR DESCRIPTION
**Pull request by dpoltavets**
Closes: #2331.

---
This change is [Reviewable](https://reviewable.io/reviews/telstra/open-kilda/2738)

### CODE REVIEW

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopology.java, line 159 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnIROB2B8krUuHjGWJH:-LnRkGgA3ZY7fsvQ3Bo_:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopology.java#L159)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Use `MessageTranslator.FIELD_ID_KEY` instead of `MessageTranslator.KEY_FIELD`.

PS It marked as deprecated.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopologyConfig.java, line 82 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnISMxQ3L6q8w6HhVRk:-LnRkMT29MTroXgRi1ho:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopologyConfig.java#L82)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

I would like to have swManager timeout * 1.5 here. It will create enough delivery tyme for corner cases (when swManager complete task almost in it's own timeout time).
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopologyConfig.java, line 96 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnISrol0v3b_D7mWPhU:-LnRkNxM2vJHSkYtzz5w:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopologyConfig.java#L96)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

What reason can be to disable it?
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopologyConfig.java, line 104 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnIT7Xs63mmCJBib_Q9:-LnRkPKp4dLn0Qddal8C:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopologyConfig.java#L104)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Will `0` value disable "switch sync" request? If not, then we need such "switch".
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java, line 101 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnITV7t4Wcn7soJOonn:-LnRkUQ02uX8n1dkRYXw:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java#L101)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Looks like, it will be easier to store "options" object.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/service/NetworkSwitchService.java, line 117 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnIUaWB-qDz5qebVzvn:-LnRkWap0ROurMouo5cU:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/service/NetworkSwitchService.java#L117)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

> ```
> locateControllerCreateIfAbsent
> ```

I believe you need `locateController` here. How "new" FSM can handle response on request sent by someone else?
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/sw/SwitchHandler.java, line 70 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnIVh6O2I3EcFo-L3Zs:-LnRka550AMYiDmxyYPO:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/sw/SwitchHandler.java#L70)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Don't see `STREAM_SWMANAGER_FIELDS`.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/sw/SwitchHandler.java, line 115 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnIVQFv80W3bqdgsqHE:-LnRkbUy4_feWYZzv37T:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/sw/SwitchHandler.java#L115)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

You can "merge" all command based input streams processing. Take a look at `PortHandler`.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/sw/SwitchHandler.java, line 202 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnIXGLBBFuAPiD5XXVT:-LnRkdFnD5JQ011Nf6ya:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/sw/SwitchHandler.java#L202)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

You do not include unique ID in request and you do not use key as such unique ID. As a result, the worker can react on outdated or reordered response. In both cases, it will be bad and unpredictable behaviour.

In other words, each request/response must be grouped to each other with some unique identifier. Also response consumer must react only on "actual" responses - with "actual" identifier.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/swmanager/SwitchManagerRouter.java, line 73 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnIYcjx9SJ-yKLYd6UJ:-LnRkggXCvD3twbEykkR:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/swmanager/SwitchManagerRouter.java#L73)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Where is "else"? Is it valid to have another type of messages here?
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/swmanager/SwitchManagerRouter.java, line 80 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnIYpbRBzKsQgOWRbz_:-LnRkiEy8RfQp1o1T61T:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/swmanager/SwitchManagerRouter.java#L80)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Same questions about "else" and validness.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[docs/design/network-discovery/switch-FSM.puml, line 31 at r3](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnkVT8kBSFrgiNFv3Nh:-Lnm7IrH3qEbpfvrB3VH:bqpumxt) ([raw file](https://github.com/telstra/open-kilda/blob/9d26d01619caabd5ae7c1e8cdece8ceedcdc6e3a/docs/design/network-discovery/switch-FSM.puml#L31)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Where is the "attemts" related checks?
</blockquote></details>

The diagram has been fixed as discussed offline.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopologyConfig.java, line 82 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnISMxQ3L6q8w6HhVRk:-Lnm7NkQ4GjrDDsNVt9r:bqmjwcz) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopologyConfig.java#L82)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Because these fields are related to each other, it whould be nice to have a comment message somethewere around, describing their relationships.
</blockquote></details>

Added.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopologyConfig.java, line 104 at r1](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnIT7Xs63mmCJBib_Q9:-Lnm7Rrn9tmIaVsWIJt_:bqmjwcz) ([raw file](https://github.com/telstra/open-kilda/blob/bdd2b1edc95bead3b0de35f2d3fe16c87f96f855/services/wfm/src/main/java/org/openkilda/wfm/topology/network/NetworkTopologyConfig.java#L104)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

Again, comment describing special behavior whould be nice yo have here.
</blockquote></details>

Added.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java, line 134 at r3](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnkW6ADBhPzsC6fUCL8:-Lnm7i_z7qj3ZLR3Bveq:bpm01mi) ([raw file](https://github.com/telstra/open-kilda/blob/9d26d01619caabd5ae7c1e8cdece8ceedcdc6e3a/services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java#L134)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

... why we have separate (from error) timeout handler?.. 
</blockquote></details>

It was decided to leave it as separate parts in the discussion offline.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java, line 279 at r3](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnkarPX3-EjuxvihoDU:-Lnm7sdNEV6R5PRyC2jK:bt1b19p) ([raw file](https://github.com/telstra/open-kilda/blob/9d26d01619caabd5ae7c1e8cdece8ceedcdc6e3a/services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java#L279)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

`check` is not the only action performed here. And even not the main action... - bad method name.
</blockquote></details>

Renamed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java, line 114 at r3](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnkbkO49UGCW02lp7yP:-Lnm7xiw74CjPdUcu0qe:bqmjwcz) ([raw file](https://github.com/telstra/open-kilda/blob/9d26d01619caabd5ae7c1e8cdece8ceedcdc6e3a/services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java#L114)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

It must be on FSM diagram.
</blockquote></details>

Added.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java, line 228 at r3](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnkcyCK6deJ9YxwExV-:-Lnm8-KQ943SvGFVo4XT:bt1b19p) ([raw file](https://github.com/telstra/open-kilda/blob/9d26d01619caabd5ae7c1e8cdece8ceedcdc6e3a/services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java#L228)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

`switch`suffix is excess here. Whole this class service switch operations.
</blockquote></details>

Renamed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java, line 231 at r3](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnkcdcA3Rz4pu87wNn7:-Lnm82HT2pRmrcV5M8QI:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/9d26d01619caabd5ae7c1e8cdece8ceedcdc6e3a/services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java#L231)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

??? You make internal transition using an external call?
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java, line 387 at r3](https://reviewable.io/reviews/telstra/open-kilda/2738#-Lnkbpt0-EWJctbyq6ZX:-Lnm852p31_h1R68O9UI:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/9d26d01619caabd5ae7c1e8cdece8ceedcdc6e3a/services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java#L387)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

You must not make "fuzzy" data lookup. If you save some data, you must use exactly this data and do not do something like this. How it can appear in context here(on place where you call getSpeakerData`)?
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/sw/SwitchHandler.java, line 86 at r3](https://reviewable.io/reviews/telstra/open-kilda/2738#-Lnkf7Y_ARgSgR1u1rVu:-Lnm8EomBrfryjS4nv2c:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/9d26d01619caabd5ae7c1e8cdece8ceedcdc6e3a/services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/sw/SwitchHandler.java#L86)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

It is bad place for such collection. This info is "tied" to FSM state and it's internal data. Take a look at BfdPortFsm.
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/swmanager/SwitchManagerRouter.java, line 73 at r3](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnkfumhBxZXwUMXsTFe:-Lnm8O-i80oUiJuOyQCt:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/9d26d01619caabd5ae7c1e8cdece8ceedcdc6e3a/services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/swmanager/SwitchManagerRouter.java#L73)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

??? it can be null? If it can - you need to raise exception about it. And you will need to have an "else" branch (without "if" in it!).
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/swmanager/SwitchManagerRouter.java, line 82 at r3](https://reviewable.io/reviews/telstra/open-kilda/2738#-LnkgAwb0POa-hYLO_O5:-Lnm8QRp3dGBCMIseGE4:bt2vr1b) ([raw file](https://github.com/telstra/open-kilda/blob/9d26d01619caabd5ae7c1e8cdece8ceedcdc6e3a/services/wfm/src/main/java/org/openkilda/wfm/topology/network/storm/bolt/swmanager/SwitchManagerRouter.java#L82)):*
<details><summary><i>Previously, surabujin (Dmitriy Bogun) wrote…</i></summary><blockquote>

same here!
</blockquote></details>

Fixed.

**Comment by dpoltavets** [comment]
___
*[services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java, line 463 at r6](https://reviewable.io/reviews/telstra/open-kilda/2738#-LoH45swE29j7keFCu1r:-LoJSbEP1-1Mr7BY0bqQ:b-dd2ixi) ([raw file](https://github.com/telstra/open-kilda/blob/800dd9679766146adf460907086bbcc587533480/services/wfm/src/main/java/org/openkilda/wfm/topology/network/controller/sw/SwitchFsm.java#L463)):*
<details><summary><i>Previously, nikitamarchenko (Nikita Marchenko) wrote…</i></summary><blockquote>

I don't see that transition on the diagram. Fix, please.
</blockquote></details>

This transition was added to the diagram.
